File: engine/git.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"strings"
)

// ── Git helpers ────────────────────────────────────────────────

func runGit(args []string, cwd string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = cwd
	cmd.Env = append(os.Environ(), "PATH="+extendedPath())
	out, err := cmd.Output()
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			return "", fmt.Errorf("%s", strings.TrimSpace(string(ee.Stderr)))
		}
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// ── Git handlers ───────────────────────────────────────────────

func handleGitStatus(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	porcelain, _ := runGit([]string{"status", "--porcelain"}, req.Cwd)
	branch, err := runGit([]string{"branch", "--show-current"}, req.Cwd)
	if err != nil || branch == "" {
		branch = "main"
	}
	files := []map[string]any{}
	for _, l := range strings.Split(porcelain, "\n") {
		if len(l) > 2 {
			files = append(files, map[string]any{
				"state": l[:2],
				"path":  strings.TrimSpace(l[3:]),
			})
		}
	}
	jsonResp(w, map[string]any{"branch": branch, "files": files, "raw": porcelain})
}

func handleGitLog(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	out, _ := runGit([]string{"log", "--oneline", "--decorate", "-30"}, req.Cwd)
	commits := []map[string]any{}
	for _, l := range strings.Split(out, "\n") {
		if l == "" {
			continue
		}
		parts := strings.SplitN(l, " ", 2)
		entry := map[string]any{"hash": parts[0], "message": ""}
		if len(parts) > 1 {
			entry["message"] = parts[1]
		}
		commits = append(commits, entry)
	}
	jsonResp(w, commits)
}

func handleGitLogGraph(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd   string `json:"cwd"`
		Limit *int   `json:"limit"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	n := 60
	if req.Limit != nil {
		n = *req.Limit
	}
	out, err := runGit([]string{
		"log",
		"--pretty=format:%H|%P|%D|%s|%an|%ar",
		fmt.Sprintf("-%d", n),
		"--all",
	}, req.Cwd)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error(), "commits": []any{}})
		return
	}
	commits := []map[string]any{}
	for _, l := range strings.Split(out, "\n") {
		if l == "" {
			continue
		}
		p := strings.SplitN(l, "|", 6)
		for len(p) < 6 {
			p = append(p, "")
		}
		parents := []string{}
		for _, s := range strings.Fields(p[1]) {
			if s != "" {
				parents = append(parents, s)
			}
		}
		refs := []string{}
		for _, s := range strings.Split(p[2], ",") {
			if t := strings.TrimSpace(s); t != "" {
				refs = append(refs, t)
			}
		}
		commits = append(commits, map[string]any{
			"hash": p[0], "parents": parents, "refs": refs,
			"subject": p[3], "author": p[4], "reltime": p[5],
		})
	}
	jsonResp(w, map[string]any{"success": true, "commits": commits})
}

func handleGitBranch(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	branch, err := runGit([]string{"branch", "--show-current"}, req.Cwd)
	if err != nil || branch == "" {
		branch = "main"
	}
	jsonResp(w, branch)
}

func handleGitBranches(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	out, _ := runGit([]string{"branch", "-a"}, req.Cwd)
	branches := []string{}
	for _, l := range strings.Split(out, "\n") {
		b := strings.TrimSpace(strings.TrimPrefix(l, "*"))
		b = strings.TrimSpace(b)
		// Skip empty and HEAD pointer entries like "remotes/origin/HEAD -> ..."
		if b == "" || strings.Contains(b, " -> ") {
			continue
		}
		branches = append(branches, b)
	}
	jsonResp(w, branches)
}

func handleGitCommit(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd     string `json:"cwd"`
		Message string `json:"message"`
	}
	if !decodeJSON(w, r, &req) {
		return
	}
	// Guard: refuse if nothing is staged
	staged, _ := runGit([]string{"diff", "--cached", "--name-only"}, req.Cwd)
	if strings.TrimSpace(staged) == "" {
		jsonResp(w, map[string]any{"success": false, "error": "nothing staged to commit"})
		return
	}
	out, err := runGit([]string{"commit", "-m", req.Message}, req.Cwd)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true, "output": out})
}

func handleGitStage(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd   string   `json:"cwd"`
		Files []string `json:"files"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	args := append([]string{"add", "--"}, req.Files...)
	if _, err := runGit(args, req.Cwd); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitUnstage(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd   string   `json:"cwd"`
		Files []string `json:"files"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	args := append([]string{"restore", "--staged", "--"}, req.Files...)
	if _, err := runGit(args, req.Cwd); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitCheckout(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd    string `json:"cwd"`
		Branch string `json:"branch"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	if _, err := runGit([]string{"checkout", req.Branch}, req.Cwd); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitPush(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	out, err := runGit([]string{"push"}, req.Cwd)
	if err != nil {
		// No upstream set — try setting it now
		branch, berr := runGit([]string{"branch", "--show-current"}, req.Cwd)
		if berr == nil && strings.TrimSpace(branch) != "" {
			out, err = runGit([]string{"push", "--set-upstream", "origin", strings.TrimSpace(branch)}, req.Cwd)
		}
		if err != nil {
			jsonResp(w, map[string]any{"success": false, "error": err.Error()})
			return
		}
	}
	jsonResp(w, map[string]any{"success": true, "output": out})
}

func handleGitPull(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	out, err := runGit([]string{"pull"}, req.Cwd)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true, "output": out})
}

func handleGitStash(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	if _, err := runGit([]string{"stash"}, req.Cwd); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitStashPop(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	if _, err := runGit([]string{"stash", "pop"}, req.Cwd); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitInit(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	if _, err := runGit([]string{"init"}, req.Cwd); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitDiscard(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd  string `json:"cwd"`
		File string `json:"file"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	// Check if file is untracked — git restore won't delete it, need clean
	st, _ := runGit([]string{"status", "--porcelain", "--", req.File}, req.Cwd)
	st = strings.TrimSpace(st)
	if strings.HasPrefix(st, "??") || strings.HasPrefix(st, "!!") {
		if _, err := runGit([]string{"clean", "-f", "--", req.File}, req.Cwd); err != nil {
			jsonResp(w, map[string]any{"success": false, "error": err.Error()})
			return
		}
	} else {
		if _, err := runGit([]string{"restore", "--", req.File}, req.Cwd); err != nil {
			if _, err2 := runGit([]string{"checkout", "--", req.File}, req.Cwd); err2 != nil {
				jsonResp(w, map[string]any{"success": false, "error": err2.Error()})
				return
			}
		}
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitDiff(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd    string `json:"cwd"`
		File   string `json:"file"`
		Staged bool   `json:"staged"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	var out string
	var err error
	if req.Staged {
		if req.File == "" {
			out, err = runGit([]string{"diff", "--cached"}, req.Cwd)
		} else {
			out, err = runGit([]string{"diff", "--cached", "--", req.File}, req.Cwd)
		}
	} else if req.File == "" {
		out, err = runGit([]string{"diff", "HEAD"}, req.Cwd)
		if err != nil {
			out, err = runGit([]string{"diff"}, req.Cwd)
		}
	} else {
		out, err = runGit([]string{"diff", "HEAD", "--", req.File}, req.Cwd)
		if err != nil {
			out, err = runGit([]string{"diff", "--", req.File}, req.Cwd)
		}
	}
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error(), "diff": ""})
		return
	}
	jsonResp(w, map[string]any{"success": true, "diff": out})
}

func handleGitBlame(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd  string `json:"cwd"`
		File string `json:"file"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	out, err := runGit([]string{"blame", "--line-porcelain", req.File}, req.Cwd)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error(), "lines": []any{}})
		return
	}
	jsonResp(w, map[string]any{"success": true, "raw": out})
}

func handleGitStashList(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	out, _ := runGit([]string{"stash", "list", "--pretty=format:%gd|%s|%cr"}, req.Cwd)
	stashes := []map[string]any{}
	for _, l := range strings.Split(out, "\n") {
		if l == "" {
			continue
		}
		p := strings.SplitN(l, "|", 3)
		for len(p) < 3 {
			p = append(p, "")
		}
		stashes = append(stashes, map[string]any{"ref": p[0], "message": p[1], "date": p[2]})
	}
	jsonResp(w, map[string]any{"success": true, "stashes": stashes})
}

func handleGitCreateBranch(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd    string `json:"cwd"`
		Branch string `json:"branch"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	if _, err := runGit([]string{"checkout", "-b", req.Branch}, req.Cwd); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitDeleteBranch(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Cwd    string `json:"cwd"`
		Branch string `json:"branch"`
	}
	json.NewDecoder(r.Body).Decode(&req)
	if _, err := runGit([]string{"branch", "-d", req.Branch}, req.Cwd); err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true})
}

func handleGitFetch(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	out, err := runGit([]string{"fetch", "--prune", "origin"}, req.Cwd)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true, "output": out})
}

func handleGitResetSoft(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	out, err := runGit([]string{"reset", "--soft", "HEAD~1"}, req.Cwd)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "error": err.Error()})
		return
	}
	jsonResp(w, map[string]any{"success": true, "output": out})
}

func handleGitAheadBehind(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	ahead, err := runGit([]string{"rev-list", "--count", "@{u}..HEAD"}, req.Cwd)
	if err != nil {
		jsonResp(w, map[string]any{"success": true, "ahead": 0, "behind": 0, "noUpstream": true})
		return
	}
	behind, _ := runGit([]string{"rev-list", "--count", "HEAD..@{u}"}, req.Cwd)
	var aheadN, behindN int
	fmt.Sscan(strings.TrimSpace(ahead), &aheadN)
	fmt.Sscan(strings.TrimSpace(behind), &behindN)
	jsonResp(w, map[string]any{"success": true, "ahead": aheadN, "behind": behindN, "noUpstream": false})
}

func handleGitRemoteList(w http.ResponseWriter, r *http.Request) {
	var req struct{ Cwd string `json:"cwd"` }
	json.NewDecoder(r.Body).Decode(&req)
	out, err := runGit([]string{"remote", "-v"}, req.Cwd)
	if err != nil {
		jsonResp(w, map[string]any{"success": false, "remotes": []any{}, "error": err.Error()})
		return
	}
	seen := map[string]bool{}
	remotes := []map[string]any{}
	for _, l := range strings.Split(out, "\n") {
		parts := strings.Fields(l)
		if len(parts) < 3 {
			continue
		}
		name := parts[0]
		if seen[name] {
			continue
		}
		seen[name] = true
		remoteType := strings.Trim(parts[2], "()")
		remotes = append(remotes, map[string]any{"name": name, "url": parts[1], "type": remoteType})
	}
	jsonResp(w, map[string]any{"success": true, "remotes": remotes})
}

File: src/lib/android-engine/git.ts
// Android Git engine — pure JS via isomorphic-git.
// No system git binary required. Works fully offline.

import git from 'isomorphic-git'
import http from 'isomorphic-git/http/web'
import { gitFsAdapter } from './fs'

const fs = gitFsAdapter

// ── Helpers ────────────────────────────────────────────────────

function shortHash(h: string) { return h.slice(0, 7) }

async function getAuthor(cwd: string) {
  try {
    const name = await git.getConfig({ fs, dir: cwd, path: 'user.name' }) ?? 'SANCTION'
    const email = await git.getConfig({ fs, dir: cwd, path: 'user.email' }) ?? 'sanction@local'
    return { name: String(name), email: String(email) }
  } catch {
    return { name: 'SANCTION', email: 'sanction@local' }
  }
}

// ── Status ─────────────────────────────────────────────────────

export async function handleGitStatus(cwd: string) {
  try {
    const statusMatrix = await git.statusMatrix({ fs, dir: cwd })
    const files: Array<{ path: string; status: string }> = []
    for (const [filepath, head, workdir, stage] of statusMatrix) {
      let status = 'unmodified'
      if (head === 0 && workdir === 2) status = '??'           // untracked
      else if (head === 1 && workdir === 1 && stage === 0) status = 'D'  // deleted (unstaged)
      else if (head === 1 && workdir === 2 && stage === 2) status = 'M'  // modified staged
      else if (head === 1 && workdir === 2 && stage === 1) status = ' M' // modified unstaged
      else if (head === 0 && workdir === 2 && stage === 2) status = 'A'  // added staged
      else if (head === 1 && workdir === 0 && stage === 1) status = ' D' // deleted unstaged
      if (status !== 'unmodified') files.push({ path: filepath, status })
    }
    const branch = await git.currentBranch({ fs, dir: cwd }) ?? 'HEAD'
    return { branch, files, clean: files.length === 0 }
  } catch (e: any) {
    return { branch: 'unknown', files: [], clean: true, error: e?.message }
  }
}

// ── Log ────────────────────────────────────────────────────────

export async function handleGitLog(cwd: string) {
  try {
    const commits = await git.log({ fs, dir: cwd, depth: 30 })
    return commits.map(c => ({
      hash: c.oid,
      short: shortHash(c.oid),
      message: c.commit.message.trim(),
      author: c.commit.author.name,
      date: new Date(c.commit.author.timestamp * 1000).toISOString(),
    }))
  } catch {
    return []
  }
}

// ── Log graph (simplified) ─────────────────────────────────────

export async function handleGitLogGraph(cwd: string, limit = 20) {
  try {
    const commits = await git.log({ fs, dir: cwd, depth: limit })
    return commits.map((c, i) => ({
      hash: c.oid,
      short: shortHash(c.oid),
      message: c.commit.message.trim(),
      author: c.commit.author.name,
      date: new Date(c.commit.author.timestamp * 1000).toISOString(),
      parents: c.commit.parent,
      refs: i === 0 ? ['HEAD'] : [],
    }))
  } catch {
    return []
  }
}

// ── Branch ─────────────────────────────────────────────────────

export async function handleGitBranch(cwd: string) {
  try {
    return { current: await git.currentBranch({ fs, dir: cwd }) ?? 'HEAD' }
  } catch {
    return { current: 'unknown' }
  }
}

export async function handleGitBranches(cwd: string) {
  try {
    const local = await git.listBranches({ fs, dir: cwd })
    const current = await git.currentBranch({ fs, dir: cwd }) ?? 'HEAD'
    return { branches: local.map(b => ({ name: b, current: b === current, remote: false })) }
  } catch {
    return { branches: [] }
  }
}

// ── Stage / Unstage ────────────────────────────────────────────

export async function handleGitStage(cwd: string, files: string[]) {
  try {
    for (const f of files) {
      await git.add({ fs, dir: cwd, filepath: f })
    }
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleGitUnstage(cwd: string, files: string[]) {
  try {
    for (const f of files) {
      await git.resetIndex({ fs, dir: cwd, filepath: f })
    }
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Commit ─────────────────────────────────────────────────────

export async function handleGitCommit(cwd: string, message: string) {
  try {
    const author = await getAuthor(cwd)
    const sha = await git.commit({ fs, dir: cwd, message, author })
    return { success: true, sha }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Checkout ───────────────────────────────────────────────────

export async function handleGitCheckout(cwd: string, branch: string) {
  try {
    await git.checkout({ fs, dir: cwd, ref: branch })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Push ───────────────────────────────────────────────────────

export async function handleGitPush(cwd: string) {
  try {
    const remote = await git.getConfig({ fs, dir: cwd, path: 'remote.origin.url' })
    if (!remote) return { success: false, error: 'No remote origin configured' }
    await git.push({ fs, dir: cwd, http, remote: 'origin' })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Pull ───────────────────────────────────────────────────────

export async function handleGitPull(cwd: string) {
  try {
    const author = await getAuthor(cwd)
    await git.pull({ fs, dir: cwd, http, author })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Init ───────────────────────────────────────────────────────

export async function handleGitInit(cwd: string) {
  try {
    await git.init({ fs, dir: cwd })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Discard (restore file to HEAD) ────────────────────────────

export async function handleGitDiscard(cwd: string, file: string) {
  try {
    await git.checkout({ fs, dir: cwd, filepaths: [file] })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Diff (simplified: show current content vs HEAD) ────────────

export async function handleGitDiff(cwd: string, file: string, staged = false) {
  try {
    let headContent = ''
    try {
      const blob = await git.readBlob({ fs, dir: cwd, oid: 'HEAD', filepath: file })
      headContent = new TextDecoder().decode(blob.blob)
    } catch { /* new file */ }
    const { readTextFile } = await import('@tauri-apps/plugin-fs')
    const workContent = await readTextFile(`${cwd}/${file}`)
    return { diff: `--- HEAD/${file}\n+++ working/${file}\n${headContent !== workContent ? 'File changed' : 'No changes'}`, headContent, workContent }
  } catch (e: any) {
    return { diff: '', error: e?.message }
  }
}

// ── Blame (simplified) ────────────────────────────────────────

export async function handleGitBlame(cwd: string, file: string) {
  try {
    const commits = await git.log({ fs, dir: cwd, depth: 50, filepath: file })
    return { blame: commits.map(c => ({
      hash: c.oid,
      author: c.commit.author.name,
      date: new Date(c.commit.author.timestamp * 1000).toISOString(),
      message: c.commit.message.trim(),
    }))}
  } catch (e: any) {
    return { blame: [], error: e?.message }
  }
}

// ── Stash (not natively supported by isomorphic-git — emulate) ─

const STASH_KEY = 'sanction-git-stash'

export async function handleGitStash(cwd: string) {
  try {
    const statusResult = await handleGitStatus(cwd)
    const { readTextFile } = await import('@tauri-apps/plugin-fs')
    const saved: Record<string, string> = {}
    for (const f of statusResult.files) {
      try { saved[f.path] = await readTextFile(`${cwd}/${f.path}`) } catch {}
      await git.checkout({ fs, dir: cwd, filepaths: [f.path] }).catch(() => {})
    }
    const existing = JSON.parse(localStorage.getItem(STASH_KEY) ?? '[]')
    existing.push({ cwd, files: saved, date: new Date().toISOString() })
    localStorage.setItem(STASH_KEY, JSON.stringify(existing))
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleGitStashPop(cwd: string) {
  try {
    const { writeTextFile } = await import('@tauri-apps/plugin-fs')
    const stashes = JSON.parse(localStorage.getItem(STASH_KEY) ?? '[]')
    const idx = stashes.findLastIndex?.((s: any) => s.cwd === cwd) ?? -1
    if (idx === -1) return { success: false, error: 'No stash found' }
    const stash = stashes.splice(idx, 1)[0]
    localStorage.setItem(STASH_KEY, JSON.stringify(stashes))
    for (const [path, content] of Object.entries(stash.files)) {
      await writeTextFile(`${cwd}/${path}`, content as string)
    }
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleGitStashList(cwd: string) {
  try {
    const stashes = JSON.parse(localStorage.getItem(STASH_KEY) ?? '[]')
    return { stashes: stashes.filter((s: any) => s.cwd === cwd) }
  } catch {
    return { stashes: [] }
  }
}

// ── Create/delete branch ───────────────────────────────────────

export async function handleGitCreateBranch(cwd: string, branch: string) {
  try {
    await git.branch({ fs, dir: cwd, ref: branch })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

export async function handleGitDeleteBranch(cwd: string, branch: string) {
  try {
    await git.deleteBranch({ fs, dir: cwd, ref: branch })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Fetch ──────────────────────────────────────────────────────

export async function handleGitFetch(cwd: string) {
  try {
    await git.fetch({ fs, dir: cwd, http })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Remote list ────────────────────────────────────────────────

export async function handleGitRemoteList(cwd: string) {
  try {
    const remotes = await git.listRemotes({ fs, dir: cwd })
    return { remotes }
  } catch {
    return { remotes: [] }
  }
}

// ── Reset soft (move HEAD back one commit) ─────────────────────

export async function handleGitResetSoft(cwd: string) {
  try {
    const commits = await git.log({ fs, dir: cwd, depth: 2 })
    if (commits.length < 2) return { success: false, error: 'No previous commit' }
    await git.writeRef({ fs, dir: cwd, ref: 'HEAD', value: commits[1].oid, force: true })
    return { success: true }
  } catch (e: any) {
    return { success: false, error: e?.message }
  }
}

// ── Ahead/behind ───────────────────────────────────────────────

export async function handleGitAheadBehind(cwd: string) {
  // isomorphic-git doesn't have a direct ahead/behind API; return defaults
  return { ahead: 0, behind: 0 }
}
